File: java/algorithms/sort/ShellSort.java
package sort;

import java.util.Arrays;

import org.junit.Test;

public class ShellSort {
	
	public void swap (int[] arr, int posA, int posB) {
		int swap = arr[posA];
		arr[posA] = arr[posB];
		arr[posB] = swap;
	}

	public void sort(int[] arr) {
		int len = arr.length;
		for (int gap = len / 2; gap > 0; gap /= 2) {
			for (int i = 0; i < gap; i++) {
				for (int j = i + gap; j < len; j += gap) {
					int k = j - gap;
					if (arr[j] < arr[k]) {
						int temp = arr[j];
						while (k >= 0 && arr[k] > temp) {
							arr[k + gap] = arr[k];
							k -= gap;
						}
						arr[k + gap] = temp;
					}
				}
			}
		}
	}

	@Test
	public void testSort() {
		int[] arr = new int[] { 49, 38, 65, 97, 26, 13, 27, 49, 55, 4, 0 };
		sort(arr);
		System.out.println(Arrays.toString(arr));
	}

	public void sort2(int[] arr) {
		int len = arr.length;
		for (int gap = len / 2; gap > 0; gap /= 2) {
			for (int j = gap; j < len; j++) {
				int k = j - gap;
				if (arr[j] < arr[k]) {
					int temp = arr[j];
					while (k >= 0 && arr[k] > temp) {
						arr[k + gap] = arr[k];
						k -= gap;
					}
					arr[k + gap] = temp;
				}
			}
		}
	}

	@Test
	public void testSort2() {
		int[] arr = new int[] { 49, 38, 65, 97, 26, 13, 27, 49, 55, 4, 0 };
		sort2(arr);
		System.out.println(Arrays.toString(arr));
	}
	
	public void sort3(int[] arr) {
		int len = arr.length;
		for(int gap = len / 2; gap > 0; gap /= 2) {
			for(int i = gap; i < len; i++) {
				for (int j = i - gap; j >= 0 && arr[j] > arr[j + gap]; j -= gap) {
					swap(arr, j, j + gap);
				}
			}
		}
	}
	
	@Test
	public void testSort3() {
		int[] arr = new int[] { 49, 38, 65, 97, 26, 13, 27, 49, 55, 4, 0 };
		sort3(arr);
		System.out.println(Arrays.toString(arr));
	}
}

File: golang/sort/shellsort.go
package sort

/*
	1 循环：选gap为一半长度
	2 循环：第一段gap
	3 循环gap间隔序列
	4 选择排序
*/
func ShellSort(arr []int) {
	length := len(arr)
	for gap := length / 2; gap > 0; gap /= 2 {
		for i := 0; i < gap; i++ {
			for j := i + gap; j < length; j += gap {
				k := j - gap
				if arr[k] > arr[j] {
					temp := arr[j]
					for k >= 0 && arr[k] > temp {
						arr[k+gap] = arr[k]
						k -= gap
					}
					arr[k+gap] = temp
				}
			}
		}
	}
}

/*
	1 循环：选gap为一半长度
	2 循环：第一个gap之后每个元素为j
	3 j向前的gap间隔序列做选择排序
 */
func ShellSort2(arr []int) {
	length := len(arr)
	for gap := length / 2; gap > 0; gap /= 2 {
		for j := gap; j < length; j++ {
			k := j - gap
			if arr[k] > arr[j] {
				temp := arr[j]
				for k >= 0 && arr[k] > temp {
					arr[k+gap] = arr[k]
					k -= gap
				}
				arr[k+gap] = temp
			}
		}
	}
}

/*
	..
	最后用冒泡排序
*/
func ShellSort3(arr []int) {
	length := len(arr)
	for gap := length / 2; gap > 0; gap /= 2 {
		for i := gap; i < length; i++ {
			for j := i - gap; j >= 0 && arr[j] > arr[j+gap]; j -= gap {
				arr[j], arr[j+gap] = arr[j+gap], arr[j]
			}
		}
	}
}
